pkg/ccl/streamingccl/streamproducer/event_stream.go

```diff
@@ -332,42 +342,58 @@ func (p *checkpointPacer) shouldCheckpoint(
 	return false
 }
 
-// Trim the received SST to only contain data within the boundaries of spans in the partition
-// spec, and execute the specified operation on each roachpb.KeyValue in the trimmed SSTable.
-func (s *eventStream) trimSST(
-	sst *roachpb.RangeFeedSSTable, op func(keyVal roachpb.KeyValue),
+// Extract the received SST to only contain data within the boundaries of spans in the partition
+// spec, execute the specified operation on each MVCC key value in the trimmed SSTable. and
+// execute the specified operation on each MVCCRangeKey value in the trimmed SSTable.
+func (s *eventStream) extractSST(
+	sst *roachpb.RangeFeedSSTable,
+	mvccKeyValOp func(key storage.MVCCKey, val []byte),
+	mvccRangeKeyOp func(rangeKeyVal storage.MVCCRangeKey),
 ) error {
-	iter, err := storage.NewMemSSTIterator(sst.Data, true)
+	iter, err := storage.NewPebbleMemSSTIterator(sst.Data, true,
+		storage.IterOptions{
+			KeyTypes:   storage.IterKeyTypePointsAndRanges,
+			LowerBound: sst.Span.Key,
+			UpperBound: sst.Span.EndKey,
+		})
 	if err != nil {
 		return err
 	}
 	defer iter.Close()
 
-	copyKV := func(mvccKey storage.MVCCKey, value []byte) roachpb.KeyValue {
-		keyCopy := make([]byte, len(mvccKey.Key))
-		copy(keyCopy, mvccKey.Key)
-		valueCopy := make([]byte, len(value))
-		copy(valueCopy, value)
-		return roachpb.KeyValue{
-			Key:   keyCopy,
-			Value: roachpb.Value{RawBytes: valueCopy, Timestamp: mvccKey.Timestamp},
+	for iter.SeekGE(storage.MVCCKey{Key: sst.RegisteredSpan.Key}); ; iter.Next() {
+		if valid, err := iter.Valid(); err != nil {
+			return err
+		} else if !valid {
+			break
 		}
-	}
-
-	return s.subscribedSpans.ForEach(func(span roachpb.Span) error {
-		iter.SeekGE(storage.MVCCKey{Key: span.Key})
-		endKey := storage.MVCCKey{Key: span.EndKey}
-		for ; ; iter.Next() {
-			if ok, err := iter.Valid(); err != nil || !ok {
-				return err
-			}
-			if !iter.UnsafeKey().Less(endKey) { // passed the span boundary
-				break
+		if !iter.UnsafeKey().Less(storage.MVCCKey{Key: sst.RegisteredSpan.EndKey}) { // passed the span boundary
+			break
+		}
+		if hasPoint, hasRange := iter.HasPointAndRange(); hasPoint {
+			mvccKeyValOp(iter.Key(), iter.Value())
+		} else if hasRange {
```

Comment:
As mentioned in another comment, let's iterate over point keys and range keys separately, unless we need to emit them interleaved in order (in which case we need to detect changes to `RangeBounds().Key`).